Add e4s tags

shahzebsiddiqui opened the pull request:
@sameershende @eugeneswalker 
This PR will add the e4s tags, the `ecp` tag is already registered with some packages so we just adding it assuming these packages are part of E4S and ECP. I used this [spack.yaml](https://github.com/E4S-Project/e4s/blob/master/spack.yaml) as a reference for identifying packages that are part of E4S. If this makes sense by all means we can start tagging the packages with `e4s`. This can be useful with `spack find -t e4s` to filter packages from non-e4s.

```
bash-3.2$ spack list -t e4s
==> 8 packages.
adios  adios2  aml  amrex  arborx  argobots  ascent  axom
```

```
bash-3.2$ spack list -t ecp
==> 29 packages.
adios   aml       amrex   argobots  axl   chombo            er    kvtree  latte      nek5000  nektools  parsplice  rankstr  shuffile  veloc
adios2  amr-wind  arborx  ascent    axom  ecp-data-vis-sdk  filo  lammps  nalu-wind  nekrs    nwchem    qmcpack    redset   spath

```

shahzebsiddiqui committed:
adding ecp and e4s tags to adios, adios2, aml, amrex, arborx, argobots, ascent, axom package

eugeneswalker commented:
Seems like a good idea!

sethrj commented:
@shahzebsiddiqui Should this be marked as a draft until all the packages are submitted? It sounds like a good idea to me but this PR is incomplete.

shahzebsiddiqui commented:
> @shahzebsiddiqui Should this be marked as a draft until all the packages are submitted? It sounds like a good idea to me but this PR is incomplete.

Well i didn't intend on updating all the E4S packages with the tags in this PR so i think this PR could be considered complete unless there is an objection. I would think that it should be maintainers adding appropriate tags, but the use of `e4s` tag helps us facility staff that deploy E4S pinpoint E4S from non-e4s packages. I think @eugeneswalker can help track state of E4S packages and update these tags as needed.

shahzebsiddiqui commented:
@sethrj if you would like me to do all the e4s packages I can do that it would update alot more files. I think @eugenewalker knows when a package becomes E4S.

sethrj commented:
I think it would be good to get at least all the e4s codes tagged. Better to update them all at once IMHO.

shahzebsiddiqui committed:
adding ecp and e4s tags to most e4s products according to https://github.com/E4S-Project/e4s/blob/master/spack.yaml
manifest file provided by E4S team

shahzebsiddiqui committed:
Merge branch 'develop' into e4s_tags

shahzebsiddiqui commented:
@sethrj sorry i just got around to doing this task. I based these changes off E4S 21.05 spack.yaml https://github.com/E4S-Project/e4s/blob/master/spack.yaml provided by @eugeneswalker . There were some packages in this list that didn't make sense why they were in e4s. Such as `gmp`? I think `py-*` extensions like `py-jupyterhub` doesn't make sense either but i did it anyways.  

Anyhow i did most of them please review them. I think i will leave this up to maintainers or E4S team to add tags when packages become E4S products.

alalazo commented:
@shahzebsiddiqui CI is not passing, due to style and supported Python versions (here it seems you use constructs somewhere that are for Python 3.6 or later)

sethrj commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -29,6 +28,8 @@ class Gasnet(Package):
 
     maintainers = ['PHHargrove', 'bonachea']
 
+    tags: ['ecp', 'e4s']
```
Tricky: `:` should be ` =`

shahzebsiddiqui replied:
thanks ill fix that.

shahzebsiddiqui replied:
this is fixed now in https://github.com/spack/spack/pull/23382/commits/a482c47ca5903eab15f28efb83b032ad57cb8d6c

sethrj reviewed (changes requested):
Those `vermin` results sure don't make it easy to find the problem file. I've tagged the error for you.

The only quibble I have is that while all of these are `e4s` specced packages, not all of them are exactly ECP software: for example, ECP doesn't fund qt or gmp development. I think the `ecp` tag would be more useful if it *only* applies to packages that are directly funded by the project. Maybe @eugeneswalker could weigh in there.

shahzebsiddiqui committed:
fix flake8 issues
typoe in tags declaration in gasnet package

shahzebsiddiqui commented:
> @shahzebsiddiqui CI is not passing, due to style and supported Python versions (here it seems you use constructs somewhere that are for Python 3.6 or later)

@alalazo flake8 should be addressed now i tested this locally will check the CI job.

tldahlgren commented on var/spack/repos/builtin/packages/argobots/package.py:
```diff
@@ -20,6 +20,8 @@ class Argobots(AutotoolsPackage):
     git      = "https://github.com/pmodels/argobots.git"
     maintainers = ['shintaro-iwasaki']
 
+    tags = ['ecp', 'e4s']
```
🤔 Is this actually an ECP project?

sethrj replied:
Nor is qt, but they seem to be explicit packages in e4s... I agree the ecp tag should be dropped if the code development isn't at least partly funded by ecp.

tldahlgren replied:
It appears in slides BUT not listed as an ST product.

shahzebsiddiqui replied:
okay `ecp` tag removed from argobots

tldahlgren commented on var/spack/repos/builtin/packages/axom/package.py:
```diff
@@ -37,6 +37,8 @@ class Axom(CachedCMakePackage, CudaPackage):
     homepage = "https://github.com/LLNL/axom"
     git      = "https://github.com/LLNL/axom.git"
 
+    tags = ['ecp', 'e4s']
```
Note this appears in slides but is not currently listed as an ST product.

eugeneswalker replied:
Definitely part of E4S. Not sure about ECP

shahzebsiddiqui replied:
okay i removed `ecp` tag from axom package

shahzebsiddiqui committed:
remove ecp tag for axom package

shahzebsiddiqui committed:
remove ecp tag from argobots

shahzebsiddiqui commented:
Please feel free to comment on all the packages that are part of `e4s` and or `ecp`. For `e4s` it was pretty straightforward. I just referenced this https://github.com/E4S-Project/e4s project and see its `spack.yaml` though i am not sure if some packages were missed from older versions of E4S.

The point of this PR is to allow us to run `spack list -t e4s` or even `spack find -t e4s`  in a deployment to see the root specs it helps to clean up the noise with all the dependent packages shown in `spack find`. This could potentially be useful for @wspear who is trying to make sense of the packages output I provide him such as this one for E4S 21.02 https://github.com/spack/spack-configs/blob/main/NERSC/cori/e4s-21.02/e4s-21.02.txt and process this into some dashboard. We could potentially run the same command with `-t <e4s>` to filter by e4s tags if we want to be explicit. Though `spack find` is still good enough as it reports everything.

sethrj commented:
Unfortunately we can't merge this now because of https://github.com/spack/spack/pull/23947 😭 it touches too many packages, and it's not fair to make you go back through and edit them all by hand.  It looks like https://github.com/spack/spack/pull/24071 provides a potential pathway to an automated solution (@cosmicexplorer).

I *do* think you should remove the `ecp` tag from all the packages on this PR that you don't explicitly know are funded by ECP; since you're picking up at least a couple that definitely aren't ECP, making the tag pretty useless.

shahzebsiddiqui commented:
@sethrj Do you want me to remove the `ecp` tag from existing packages before this PR i think there was about a dozen or so packages that have this. Sure i can remove the `ecp` tags from the ones I specified so far. Ill get back to you by tomorrow

sethrj commented:
> @sethrj Do you want me to remove the `ecp` tag from existing packages before this PR i think there was about a dozen or so packages that have this. Sure i can remove the `ecp` tags from the ones I specified so far.

I would say to assume that people who individually tagged ecp did so for a reason besides "bundled with e4s" so just remove the ecp tags from the ones you're adding to this pr.

shahzebsiddiqui committed:
remove ecp tag from select products that are not funded by Exascale Computing Project (ECP)

shahzebsiddiqui committed:
Merge branch 'develop' into e4s_tags

shahzebsiddiqui commented:
I have removed the `ecp` tag from products that i know dont belong to ECP. I used one of our dashboards in our ECP confluence https://confluence.exascaleproject.org/display/HISD/SI+Build+and+Test+Dashboard to keep track of this @tldahlgren is one of the owners for this dashboard along with Ryan Adamson. I reviewed all of the specs i updated and removed package that were not in the list. This will address several packages that are part of ECP project but didnt have a `ecp` tag. This way we get both of these tags registered in one shot. I am sure i missed some packages it is up to maintainers to add these tags which should be relatively easy. Perhaps one of the E4S teams @eugeneswalker @wspear can update these tags when products are part of E4S with `e4s` tag. 

Not sure if that maintainer bot @adamjstewart created can help ping all the maintainers now to confirm. I think most products covered in this PR are part of ECP and E4S.

adamjstewart commented:
The way the maintainer bot is designed, it won't ping people for review unless it's a new PR being created. This prevents the bot from pinging the exact same people every time a new commit is added. @tgamblin is working on a plugin that will allow you to manually run the bot by commenting with @spackbot, but I don't think it's ready yet.

sethrj commented:
@shahzebsiddiqui Thankfully @tgamblin just merged a fix for *all* the import orderings, which means you should be able to merge/rebase and get a passing CI. 😅

adamjstewart commented:
Not yet, #24695 still hasn't been merged.

sethrj commented:
> Not yet, #24695 still hasn't been merged.

Whoops, I must've mistaken the "tgamblin push" on my homepage: I assumed it was to `develop` but it probably shows updates for feature branches too 🙄

shahzebsiddiqui committed:
Merge branch 'develop' into e4s_tags

shahzebsiddiqui committed:
Merge branch 'develop' into e4s_tags

shahzebsiddiqui commented:
I rebased my branch and fixed all the merge conflicts. Looks like the import errors in style are still going to be there https://github.com/spack/spack/pull/23382/checks?check_run_id=3010350761 until we get this merged. Let me know once this is merged and i can rebase again

sethrj commented:
Ugh. Sorry about all the hassle. 😞

shahzebsiddiqui commented:
Okay i wanted to point out we now have some direction onto E4S products. We will start to document this in the e4s project at https://github.com/E4S-Project/e4s which will be the source of truth. There is an open PR: https://github.com/E4S-Project/e4s/pull/19 that i am working on which requires review from E4S team. Once this is merged, I can update this MR with the list of `e4s` tags. As per `ecp` tags ill need to check on that list and get back to you with an update

shahzebsiddiqui committed:
Merge branch 'develop' into e4s_tags

shahzebsiddiqui committed:
Merge branch 'e4s_tags' of github.com:shahzebsiddiqui/spack into e4s_tags

shahzebsiddiqui committed:
remove 'ecp' tags from all packages
add 'e4s' tags for all packages according to https://github.com/E4S-Project/e4s/blob/master/E4S_Products.md

shahzebsiddiqui commented:
@tldahlgren 
cc: @sameershende 

I have finally got all the packages updated. I have removed `ecp` tag from all packages that I added. I have updated all packages with `e4s` tag according to https://github.com/E4S-Project/e4s/blob/master/E4S_Products.md which is the official list for e4s products. This list may change over time in that case E4S team can go in and update the spack package with the `e4s` tag. Currently we have 102 packages with e4s tag. There are total of 88 products in the E4S product table however some products have more than one spack package such as `darshan`, `flux`, `llvm`, `superlu`, etc which is why we have this difference.

Please review this and merge this when its ready.

```
bash-3.2$ spack list -t e4s
==> 102 packages.
adios     bolt             dyninst      globalarrays    libnrm         mpich             papyrus          py-cinemasci    scr           tasmanian  warpx
adios2    cabana           faodel       gotcha          libquo         mpifileutils      parallel-netcdf  py-jupyterhub   slate         tau        zfp
aml       caliper          flecsi       hdf5            llvm           netlib-scalapack  paraview         py-libensemble  slepc         trilinos
amrex     chai             flit         heffte          llvm-doe       ninja             parsec           py-petsc4py     stc           umap
aomp      charliecloud     flux-core    hpctoolkit      loki           nrm               pdt              py-warpx        strumpack     umpire
arborx    conduit          flux-sched   hpx             magma          nvhpc             petsc            qt              sundials      unifyfs
archer    darshan-runtime  fortrilinos  hypre           mercury        omega-h           phist            qthreads        superlu       upcxx
argobots  darshan-util     gasnet       kokkos          metall         openmpi           plasma           qwt             superlu-dist  variorum
ascent    datatransferkit  geopm        kokkos-kernels  mfem           openpmd-api       precice          raja            swig          veloc
axom      dealii           ginkgo       legion          mpark-variant  papi              pumi             rempi           sz            vtk-m
```

bonachea commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -17,7 +17,6 @@ class Gasnet(Package):
        primarily intended as a compilation target and for use by runtime library
        writers (as opposed to end users), and the primary goals are high
        performance, interface portability, and expressiveness.
-
```
Why was this whitespace / paragraph break removed from our package description?

This is here to emphasize the notice below.

shahzebsiddiqui replied:
i added the new line in https://github.com/spack/spack/pull/23382/commits/43013f54f71db4fb2d517225cad0b421564841cc.

The newline doesn't show up in `spack info` so it doesn't make a difference apart from readability. 

```
bash-3.2$ spack info gasnet
Package:   gasnet

Description:
    GASNet is a language-independent, networking middleware layer that
    provides network-independent, high-performance communication primitives
    including Remote Memory Access (RMA) and Active Messages (AM). It has
    been used to implement parallel programming models and libraries such as
    UPC, UPC++, Co-Array Fortran, Legion, Chapel, and many others. The
    interface is primarily intended as a compilation target and for use by
    runtime library writers (as opposed to end users), and the primary goals
    are high performance, interface portability, and expressiveness.
    ***NOTICE***: The GASNet library built by this Spack package is ONLY
    intended for unit-testing purposes, and is generally UNSUITABLE FOR
    PRODUCTION USE. The RECOMMENDED way to build GASNet is as an embedded
    library as configured by the higher-level client runtime package (UPC++,
    Legion, etc), including system-specific configuration.
```

bonachea replied:
> The newline doesn't show up in spack info so it doesn't make a difference apart from readability.

This is a _defect_ in the current implementation of `spack info`, reported as issue #23831 that will hopefully be corrected soon.

Thanks for restoring the formatting that will hopefully be rendered correctly in an upcoming release.

tldahlgren commented on var/spack/repos/builtin/packages/mfem/package.py:
```diff
@@ -13,7 +13,7 @@
     """Free, lightweight, scalable C++ library for finite element methods."""
 
-    tags = ['fem', 'finite-elements', 'high-order', 'amr', 'hpc', 'radiuss']
+    tags = ['fem', 'finite-elements', 'high-order', 'amr', 'hpc', 'e4s']
```
```suggestion
    tags = ['fem', 'finite-elements', 'high-order', 'amr', 'hpc', 'radiuss', 'e4s']
```

tldahlgren replied:
Don't drop the `radiuss` tag.

shahzebsiddiqui replied:
that was my mistake. Will add this in

tldahlgren commented on var/spack/repos/builtin/packages/mpich/package.py:
```diff
@@ -19,9 +19,9 @@ class Mpich(AutotoolsPackage):
     list_url = "https://www.mpich.org/static/downloads/"
     list_depth = 1
-
+    
```
Style issue.

```suggestion

```

shahzebsiddiqui replied:
good catch. I will fix this

shahzebsiddiqui replied:
i fixed all the style issues reported in CI check. There were a few more packages that had same issue.

tldahlgren commented on var/spack/repos/builtin/packages/sz/package.py:
```diff
@@ -14,6 +14,14 @@ class Sz(CMakePackage):
 
+    tags = ['e4s']
+
+    def url_for_version(self, version):
+        """provide url to ensure that download counting via github releases
+        works accurately"""
+        url = "https://github.com/szcompressor/SZ/releases/download/v{0}/SZ-{0}.tar.gz"
+        return url.format(version)
+
```
This appears to be out of scope for this PR.  Was its addition here intentional?

shahzebsiddiqui replied:
that was not intention. That was me rebasing my branch and resulted in merge conflict. anyhow i checked the sz package https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/sz/package.py and it doesn't have this section so i removed it

sethrj commented:
@shahzebsiddiqui Looks like some rebases might have gone horribly awry.

shahzebsiddiqui committed:
remove function from sz package
add new line in gasnet package description between 'notice'
fix style issue

sethrj commented on var/spack/repos/builtin/packages/mpich/package.py:
```diff
@@ -19,9 +19,9 @@ class Mpich(AutotoolsPackage):
     list_url = "https://www.mpich.org/static/downloads/"
     list_depth = 1
-
+ 
```
Whitespace error

shahzebsiddiqui replied:
all style checks have been addressed and I checked the latest CI check and i saw no errors.

sethrj reviewed (dismissed):
Looks good but I think you have a whitespace error that will cause the style check to fail.

shahzebsiddiqui committed:
fix spack style issues

sethrj reviewed (approved):
Looks good, the only changes are adding `e4s` tags. @tldahlgren this will need your re-approval before it's merged.
